src/Interpreters/HashJoin.cpp

```diff
@@ -1081,8 +1081,7 @@ class AddedColumns
 
     struct LazyOutput
     {
-        PaddedPODArray<UInt64> blocks;
-        PaddedPODArray<UInt32> row_nums;
+        PaddedPODArray<UInt64> row_refs;
```

Comment:
and why it is beneficial in the end? just because we filling in one array instead of two?

Reply:
I think it is because the fill array operation of `addFoundRowAll` is inside the `joinRightColumns` for loop, which will scan all the left table rows, and if the right table rows has a lots of matched rows, then the time complexity is very high. so I move the `for` loop of `addFoundRowAll` to the `buildOutput` to reduce the time complexity of `joinRightColumns`.   and on another hand, the old code will fill 2 array, and now it fill one.

Reply:
indeed, it could be almost fixed by the following patch: https://pastila.nl/?0169ab09/eb769dfb95b000c681651866b8c89d86#VYcmBdzNznC3fy/9hHzw3Q==
afaiu the similar problem should exist for `buildOutputFromRowRef` as well, so it makes sense to unify their implementation and use this patch in both cases

Reply:
done

Reply:
> indeed, it could be almost fixed by the following patch: https://pastila.nl/?0169ab09/eb769dfb95b000c681651866b8c89d86#VYcmBdzNznC3fy/9hHzw3Q== afaiu the similar problem should exist for `buildOutputFromRowRef` as well, so it makes sense to unify their implementation and use this patch in both cases

This patch make test case `hashjoin_with_large_output`  10% slowdown. @nickitat